src/pykeen/evaluation/restricted_evaluator.py

```diff
@@ -0,0 +1,180 @@
+"""Sampled evaluator from [teru2020]_."""
+from collections import defaultdict
+from typing import Dict, Tuple, Optional
+
+import numpy as np
+import torch
+
+from .rank_based_evaluator import RankBasedEvaluator, compute_rank_from_scores, SIDE_HEAD, SIDE_TAIL
+from ..triples import CoreTriplesFactory
+from ..triples.triples_factory import MappedTriples
+
+
+def sample_negatives(
+    valid_triples: CoreTriplesFactory,
+    all_pos: CoreTriplesFactory,
+    num_samples: int = 50,
+) -> Tuple[torch.Tensor, torch.Tensor, Dict]:
+
+    """
+    Samples num_samples _filtered_ negative entities, ie, we make sure
+    triples with sampled heads / tails do not exist in the input graph.
+    For the inductive LP setup, the input graph is supposed to be inductive_inference graph,
+    not transductive_training.
+
+    Currently, the code is not very efficient in favor of replicating the original behavior
+
+    :param valid_triples: CoreTriplesFactory
+        Triples to be evaluated
+    :param all_pos: CoreTriplesFactory
+        The inference graph on which we run the valid_triples. Needed for filtering
+    :param num_samples: int
+        Number of random entities to sample
+    """
+
+    val_triples = valid_triples.mapped_triples
+    all_pos_triples = all_pos.mapped_triples
+    num_entities = all_pos.num_entities
+
+    head_samples, tail_samples = [[] for _ in range(len(val_triples))], [[] for _ in range(len(val_triples))]
+
+    # indexing positive triples
+    head_index, tail_index = defaultdict(list), defaultdict(list)
+    negs_dict = {"head": {}, "tail": {}}
+
+    for triple in all_pos_triples:
+        h, r, t = triple[0].item(), triple[1].item(), triple[2].item()
+        tail_index[(h, r)].append(t)
+        head_index[(r, t)].append(h)
+
+    # sampling N negatives for each triple
+    for i, row in enumerate(val_triples):
+        head, rel, tail = row[0].item(), row[1].item(), row[2].item()
+
+        head_samples[i].append(head)
+        while len(head_samples[i]) < num_samples:
```

Comment:
I'm suspect of a while loop

Reply:
can't you do multiple random choices without replacement in the beginning instead

Reply:
technically it's possible - those loops do not look very appealing, I took them from the original code for the sake of replicating the envisioned behavior